lightning/src/ln/peer_handler.rs

```diff
@@ -44,6 +43,17 @@ use bitcoin::hashes::sha256::Hash as Sha256;
 use bitcoin::hashes::sha256::HashEngine as Sha256Engine;
 use bitcoin::hashes::{HashEngine, Hash};
 
+/// Handler for messages external to the LN protocol.
+pub trait CustomMessageHandler : wire::CustomMessageReader {
+	/// Called with the message type that was received and the buffer to be read.
+	/// Can return a `MessageHandlingError` if the message could not be handled.
+	fn handle_custom_message(&self, msg: Self::CustomMessage) -> Result<(), LightningError>;
+
+	/// Gets the list of pending messages which were generated by the custom message
+	/// handler, clearing the list in the process.
+	fn get_and_clear_pending_msg(&self) -> Vec<(PublicKey, Self::CustomMessage)>;
```

Comment:
Precise the `PublicKey` must be node id ?

If you want to send the same message to N peers, should the return type be instead `Vec<CustomMessage, Vec<PublicKey>>` ? More memory conservative.

Reply:
Added some doc.

Updated to `Vec<PublicKey>`. Though I would note that in terms of memory it probably means an increase for the single recipient case.

Reply:
> If you want to send the same message to N peers, should the return type be instead Vec<CustomMessage, Vec<PublicKey>> ? More memory conservative.

Strongly disagree, I don't think this is a particularly useful API - if you want to broadcast a message, getting the list of peers connected seems like a lot of work, so I'm not sure this is the right API for it anyway. If you want to unicast a message (which I believe is the common case), now you're wasting a ton of extra memory to do a full second heap allocated `Vec` just to store a single publickey.

Reply:
> so I'm not sure this is the right API for it anyway. If you want to unicast a message (which I believe is the common case),

Yeah I see where we diverge, I've more in mind multicast type of broadcast such as headers or dlc offers. Especially if you start to broadcast dlc offers per-subset such as future, options, twitter bets...I think we can revisit the API in the future, if memory starts to be a real bottleneck.

Though note as it's public API, won't be free to migrate library users.